pkg/app/piped/eventwatcher/eventwatcher.go

```diff
@@ -220,50 +235,97 @@ func (w *watcher) updateValues(ctx context.Context, repo git.Repo, repoID string
+		outDatedEvents = make([]*pipedservice.ReportEventStatusesRequest_Event, 0)
+		maxTimestamp   int64
+	)
 	for _, e := range eventCfgs {
-		latestEvent, ok := w.eventStore.GetLatest(ctx, e.Name, e.Labels)
-		if !ok {
+		notHandledEvents := w.eventLister.ListNotHandled(e.Name, e.Labels, milestone.(int64)+1, numToMakeOutdated)
+		if len(notHandledEvents) == 0 {
 			continue
 		}
-		if latestEvent.IsHandled() {
-			continue
+		if len(notHandledEvents) > 1 {
+			// Events other than the latest will be OUTDATED.
+			for i := 1; i < len(notHandledEvents[1:]); i++ {
+				outDatedEvents = append(outDatedEvents, &pipedservice.ReportEventStatusesRequest_Event{
+					Id:                notHandledEvents[i].Id,
+					Status:            model.EventStatus_EVENT_OUTDATED,
+					StatusDescription: fmt.Sprintf("The new event %q has been created", notHandledEvents[i].Id),
+				})
+			}
+		}
+
+		latestEvent := notHandledEvents[0]
+		if firstRead {
+			// Avoid handling event that isn't the latest.
+			resp, err := w.apiClient.GetLatestEvent(ctx, &pipedservice.GetLatestEventRequest{
+				Name:   e.Name,
+				Labels: e.Labels,
+			})
+			if err != nil {
+				return fmt.Errorf("failed to get the latest event: %w", err)
+			}
+			if resp.Event.CreatedAt > latestEvent.CreatedAt {
+				continue
+			}
 		}
-		le := *latestEvent
 		if err := w.commitFiles(ctx, latestEvent.Data, e, tmpRepo, commitMsg); err != nil {
 			w.logger.Error("failed to commit outdated files", zap.Error(err))
-			le.Status = model.EventStatus_EVENT_FAILURE
-			le.StatusDescription = fmt.Sprintf("Failed to change files: %v", err)
-			newEvents = append(newEvents, le)
+			handledEvents = append(handledEvents, &pipedservice.ReportEventStatusesRequest_Event{
+				Id:                latestEvent.Id,
+				Status:            model.EventStatus_EVENT_FAILURE,
+				StatusDescription: fmt.Sprintf("Failed to change files: %v", err),
+			})
 			continue
 		}
-		le.Status = model.EventStatus_EVENT_SUCCESS
-		le.StatusDescription = fmt.Sprintf("Successfully updated %d files in the %q repository", len(e.Replacements), repoID)
-		newEvents = append(newEvents, le)
+		handledEvents = append(handledEvents, &pipedservice.ReportEventStatusesRequest_Event{
+			Id:                latestEvent.Id,
+			Status:            model.EventStatus_EVENT_SUCCESS,
+			StatusDescription: fmt.Sprintf("Successfully updated %d files in the %q repository", len(e.Replacements), repoID),
+		})
+		if latestEvent.CreatedAt > maxTimestamp {
+			maxTimestamp = latestEvent.CreatedAt
+		}
 	}
+	if len(outDatedEvents) > 0 {
+		if _, err := w.apiClient.ReportEventStatuses(ctx, &pipedservice.ReportEventStatusesRequest{Events: outDatedEvents}); err != nil {
+			return fmt.Errorf("failed to report event statuses: %w", err)
+		}
+		w.logger.Info(fmt.Sprintf("successfully made %d events OUTDATED", len(outDatedEvents)))
+	}
+	if len(handledEvents) == 0 {
+		return nil
+	}
+	w.milestoneMap.Store(repoID, maxTimestamp)
```

Comment:
Shouldn't we just save this value after calling function `ReportEventStatuses`?

Reply:
Right 👍